### ISSUE TITLE
Doom hangs when editing a file over ssh as root

### ISSUE THREAD

**Issue by lunik1** [hlissner/doom-emacs#4773]
**What did you expect to happen?**

Emacs not to hang while editing a file over ssh with TRAMP with elevated privileges

**What actually happened?**

After opening a remote file using TRAMP+ssh and editing it using sudo with `SPC f U`, Emacs will hang after a short - but variable - period of time. Usually within a few minutes. Emacs will not respond to any input, even `C-g`*.

*after about a dozen presses the file name in the mode line will disappear, but emacs will still not respond to any other input.

**Additional details:**

I have been able to replicate this with DOOM's default configuration.

These are, I think, relevant messages. It's hard to get much more debugging info since Emacs has stopped responding at this point.
```
Tramp: Opening connection for root@remote using sudo...
Tramp: Sending command ‘exec ssh    -e none remote’
Tramp: Waiting for prompts from remote shell...failed
Tramp: Opening connection for root@remote using sudo...failed
Cleaning up the recentf list...
File /sudo:root@remote:/etc/nixos/configuration.nix removed from the recentf list
Error during redisplay: (eval (doom-modeline-segment--buffer-info)) signaled (quit "")
```

`remote` does not accept ssh logins as `root`.

**Steps to reproduce:**
1. Open a remote file using TRAMP's ssh method
2. Elevate privileges with `SPC f U`
3. Wait, this should happen after ⪅10 minutes

The same issue occurs if using TRAMP's sudo method in conjunction with `tramp-default-proxies-alist`

**System information:**
```
SYSTEM  info       (Arch Linux . Linux 4.19.104-microsoft-standard x86_64)
        shell      /bin/zsh
        path       (~/.cargo/bin ~/bin ~/.cargo/bin/ /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /usr/games /usr/local/games /mnt/c/Program Files/Alacritty/ /mnt/c/Python39/Scripts/ /mnt/c/Python39/ /mnt/c/Python38/Scripts/ /mnt/c/Python38/ /mnt/c/Program Files (x86)/Common Files/Oracle/Java/javapath /mnt/c/Python37/Scripts/ /mnt/c/Python37/ /mnt/c/Windows/system32 /mnt/c/Windows /mnt/c/Windows/System32/Wbem /mnt/c/Windows/System32/WindowsPowerShell/v1.0/ /mnt/c/Windows/System32/OpenSSH/ /mnt/c/ProgramData/chocolatey/bin /mnt/c/ProgramData/chocolatey/lib/mpv.install/tools /mnt/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common /mnt/c/Program Files/OpenSSH-Win64 /mnt/c/Program Files/OpenJDK/jdk-12.0.1/bin /mnt/c/Program Files/OpenJDK/jdk-12.0.2/bin /mnt/c/Program Files/OpenJDK/jdk-13/bin /mnt/c/Program Files (x86)/GNU/GnuPG/pub /mnt/c/Program Files/Git/cmd /mnt/c/WINDOWS/system32 /mnt/c/WINDOWS /mnt/c/WINDOWS/System32/Wbem /mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/ /mnt/c/WINDOWS/System32/OpenSSH/ /mnt/c/Strawberry/c/bin /mnt/c/Strawberry/perl/site/bin /mnt/c/Strawberry/perl/bin /mnt/c/Program Files/Mozilla Firefox /mnt/c/Program Files (x86)/GtkSharp/2.12/bin /mnt/c/Program Files (x86)/Streamlink/bin /mnt/c/Program Files/NVIDIA Corporation/NVIDIA NvDLISR /mnt/c/Program Files/Docker/Docker/resources/bin /mnt/c/ProgramData/DockerDesktop/version-bin /mnt/c/Program Files/OpenJDK/jdk-15.0.2/bin /mnt/c/ProgramData/chocolatey/lib/gsudo/bin/ /mnt/c/Program Files/PowerShell/7/ /mnt/c/Program Files/nodejs/ /mnt/c/Users/$USER/AppData/Local/Microsoft/WindowsApps /mnt/c/tools/neovim/Neovim/bin /mnt/c/Program Files/MPC-BE x64/ /mnt/c/Program Files (x86)/FAHClient /mnt/c/Program Files/MKVToolNix /mnt/c/Users/$USER/AppData/Roaming/npm /usr/lib/jvm/default/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /usr/lib/emacs/27.1/x86_64-pc-linux-gnu)
EMACS   dir        ~/.emacs.d/
        version    (27.1 Aug 28, 2020)
        buildopts  --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --with-x-toolkit=gtk3 --with-xft --with-wide-int --with-modules --with-cairo --with-harfbuzz 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP
        traits     (gui server-running)
DOOM    dir        ~/.doom.d/
        version    (2.0.9 HEAD -> develop e4539bb56 2021-03-06 10:04:54 -0500)
        font       #<font-object -ADBO-Source Code Pro-normal-normal-normal-*-13-*-*-*-m-0-iso10646-1>
        theme      doom-one
        modules    (:completion (company +childframe) (ivy +icons +fuzzy) :ui doom doom-dashboard doom-quit fill-column hl-todo hydra indent-guides (ligatures +extra) modeline ophints (popup +all) treemacs vc-gutter vi-tilde-fringe (window-select +numbers +ace-window) workspaces zen :editor (evil +everywhere) file-templates fold format snippets :emacs (dired +ranger +icons) electric (ibuffer +icons) (undo +tree) vc :term eshell shell vterm :checkers (syntax +childframe) (spell +aspell +everywhere) grammar :tools biblio (docker +lsp) (eval +overlay) (lookup +dictionary) (lsp + peek) (magit +forge) pdf rgb (&nopath tree-sitter) :os tty :lang (cc +lsp) clojure data emacs-lisp json (julia +lsp) (latex +latexmk +lsp) markdown nix (org +noter +pretty +roam) (python +poetry +lsp) (rust +lsp) (sh +lsp) (yaml +lsp) :config literate (default +bindings +smartparens))
        packages   ((flycheck-clojure) (info-colors) (julia-formatter :recipe (:host github :repo ki-chi/julia-formatter)) (keyfreq) (lsp-julia :recipe (:host github :repo non-jedi/lsp-julia)) (lsp-latex) (magit-delta) (org-appear :recipe (:host github :repo awth13/org-appear)) (org-fragtog) (org-pandoc-import :recipe (:host github :repo tecosaur/org-pandoc-import :files (*.el filters preprocessors))) (org-ref) (org-roam-bibtex) (ox-gfm) (ox-rst) (page-break-lines) (rotate) (vdiff) (vimrc-mode) (yaml-mode :recipe (:host gitlab :repo lunik1/yaml-mode)))
        unpin      (yaml-mode)
        elpa       (n/a)
```
Also reproduced on NixOS 20.09 and unstable.

**Comment by hlissner** [comment]
Duplicate of #3909

**Comment by lunik1** [comment]
I'm not sure #3909 is exactly the same issue. Doom is responsive enough for me when editing files over ssh when not root. And the behaviour described in #3909 sounds more like lag, but in my case one Emacs hangs it will not recover.

I can see how it could be related to TRAMP re-opening a connection, but could this still be an issue if TRAMP has to re-open a connection for some other reason?

**Comment by lunik1** [comment]
`"/ssh:remote|sudo:root@remote:/etc/nixos/configuration.nix"` which seems sensible: it is not attempting to ssh as root directly. `SPC f U` does work as expected until Emacs starts to hang.

So maybe the mysterious TRAMP connection re-openings are ignoring this? In which case this should be resolved with #3909.

**Comment by lunik1** [comment]
Aha! I have found a way to trigger this manually. After opening the file as root with `SPC f U` run `tramp-cleanup-connection` on the `/sudo:root@remote` connection. Emacs hangs and I see the same Messages after jamming on `C-g` for a while.
